File: longest_substring_without_duplicates.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        hashmap = {}
        length = 0

        for r in range(len(s)):
            hashmap[s[r]] = hashmap.get(s[r], 0) + 1
            while hashmap[s[r]] > 1:
                hashmap[s[l]] -= 1
                l += 1
            length = max(length, (r - l) + 1)

        return length

File: longestSubstringWithoutDuplicates.js
class Solution {
  /**
   * @param {string} s
   * @return {number}
   */
  lengthOfLongestSubstring(s) {
    let hash = {};
    let len = 0;
    let l = 0;

    for (let r = 0; r < s.length; r++) {
      hash[s[r]] = 1 + (hash[s[r]] || 0);
      while (hash[s[r]] > 1) {
        hash[s[l]]--;
        l++;
      }
      len = Math.max(len, r - l + 1);
    }
    return len;
  }
}
